### PR TITLE
Enable pickling for LLVMDouble class

### LINKED ISSUE #212: Support save and load of LLVM Lambdify
I think we'll have to save the internal variables and the LLVM function. I have no experience with using pickle with extension types. Anybody have an idea how to do this?

cc @bjodah, @ichumuh.

### PR DESCRIPTION
**Pull request by isuruf**
Fixes #212

### CODE REVIEW

**Comment by bjodah** [comment]
You're (insanely) fast @isuruf :) -- I was just going to start looking into this.
Changing load/dump to loads/dumps was on the top of my list but you're way ahead here.
The approach looks great to me. +1 to merge when tests pass.

File: symengine/lib/symengine_wrapper.pyx [hunk 1]
```diff
@@ -4369,7 +4369,7 @@ cdef class _Lambdify(object):
     cdef vector[int] accum_out_sizes
     cdef object numpy_dtype
 
-    def __init__(self, args, *exprs, cppbool real=True, order='C', cppbool cse=False):
+    def __init__(self, args, *exprs, cppbool real=True, order='C', cppbool cse=False, cppbool load=False):
```
Comment (bjodah) [inline symengine/lib/symengine_wrapper.pyx]: maybe we should name the kwarg `_load` to signal that end-users should be touching it?

Comment (isuruf) [reply symengine/lib/symengine_wrapper.pyx]: Good suggestion. I'll change this

File: symengine/lib/symengine_wrapper.pyx [hunk 2]
```diff
@@ -4639,6 +4661,8 @@ IF HAVE_SYMENGINE_LLVM:
             addr2 = cast(<size_t>&self.lambda_double[0], c_void_p)
             return create_low_level_callable(self, addr1, addr2)
 
+    def llvm_loading_func(*args):
+        return LLVMDouble(args, _load=True)
```
Comment (bjodah) [inline symengine/lib/symengine_wrapper.pyx]: Actually I think the `_load` kwarg could go into `__cinit__` if we `cdef` the return value here:
```
def llvm_loading_func(*args):
    cdef LLVMDouble llvm_double = LLVMDouble(args, _load=True)
```

cf. e.g. [this](https://github.com/bjodah/symcxx/blob/master/symcxx/_symcxx.pyx.mako_template#L450)

**Comment by bjodah** [comment]
Somewhat unrelated: in SymEngine (the C++ library) all `Basic` instances are immutable and they have typeid, so I guess they serialize? I haven't found the function for that though, is there any such? (I'm thinking they should in principle serialize to bytestreams too?)

**Comment by isuruf** [comment]
> in SymEngine (the C++ library) all Basic instances are immutable and they have typeid, so I guess they serialize? I haven't found the function for that though, is there any such? (I'm thinking they should in principle serialize to bytestreams too?)

I know nothing about serialization in C++. Can you open an issue about it in the C++ repo, so that we can ask those who know about it?

File: symengine/lib/symengine_wrapper.pyx (hunk 2 above)
Comment (isuruf) [reply symengine/lib/symengine_wrapper.pyx]: I'm not sure I follow. Can you elaborate more? (Or a PR. :smiley:)

Comment (bjodah) [reply symengine/lib/symengine_wrapper.pyx]: It's not important (actually not sure it matters much), so let's leave it for now (and I can try to make a PR against master after this is merged if I can find a way to hide the _load kwarg).

**Comment by bjodah** [comment]
> I know nothing about serialization in C++. Can you open an issue about it in the C++ repo, so that we can ask those who know about it?

I opened https://github.com/symengine/symengine/issues/1394

**Pull request merged**